### ISSUE TITLE
Possible Bug: Front End Form validation

### ISSUE THREAD

**Issue by mikestecker** [craftcms/cms#1636]
### Description

I am in the early stages of building a front-end form and I'm having issues with submitting entry field types with empty data. The form won't save the new entry without something as a value for the field.

When I first began to try and diagnose the error I didn't have a field in my front-end form for this particular field because it was meant for me to use on an admin level in the background after a user has submitted the form. This would allow me to cross-reference an entry in another section of the site.

The form field I added to the front end for this field is `<input type="hidden" name="fields[client][]" value="">` The field `client` is a entry field to link to another section's entry.

I have tried submitting the form with the field the way above, also as `<input type="hidden" name="fields[client]" value="">`, which is how it seems to come across in the CMS POST (shows as `'client' => ''`.

I always get an error unless I put something in as a value, such as `<input type="hidden" name="fields[client][]" value="27">`. When I put in a value, the form saves the new entry without error.

Another thing of note is that I have the entry title being generated from my `clientCompanyName` field and my `dateOfVisit` field.

At the very end of this issue I will put in my entire form template code....

The error in the log is

```
craft\db\QueryAbortedException in /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/craftcms/cms/src/elements/db/ElementQuery.php:854 Stack trace:
#0 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/yiisoft/yii2/db/QueryBuilder.php(92): craft\elements\db\ElementQuery->prepare(Object(craft\db\mysql\QueryBuilder))
#1 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/yiisoft/yii2/db/Query.php(126): yii\db\QueryBuilder->build(Object(craft\elements\db\EntryQuery))
#2 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/yiisoft/yii2/db/Query.php(418): yii\db\Query->createCommand(Object(craft\db\Connection))
#3 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/craftcms/cms/src/db/Query.php(197): yii\db\Query->queryScalar('COUNT(*)', NULL)
#4 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/yiisoft/yii2/db/Query.php(319): craft\db\Query->queryScalar('COUNT(*)', NULL)
#5 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/craftcms/cms/src/elements/db/ElementQuery.php(996): yii\db\Query->count('*', NULL)
#6 [internal function]: craft\elements\db\ElementQuery->count()
#7 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/craftcms/cms/src/validators/ArrayValidator.php(115): count(Object(craft\elements\db\EntryQuery))
#8 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/yiisoft/yii2/validators/Validator.php(267): craft\validators\ArrayValidator->validateAttribute(Object(craft\elements\Entry), 'client')
#9 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/yiisoft/yii2/base/Model.php(358): yii\validators\Validator->validateAttributes(Object(craft\elements\Entry), Array)
#10 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/craftcms/cms/src/services/Elements.php(373): yii\base\Model->validate()
#11 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/craftcms/cms/src/controllers/EntriesController.php(481): craft\services\Elements->saveElement(Object(craft\elements\Entry))
#12 [internal function]: craft\controllers\EntriesController->actionSaveEntry()
#13 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#14 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/yiisoft/yii2/base/Controller.php(156): yii\base\InlineAction->runWithParams(Array)
#15 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/yiisoft/yii2/base/Module.php(523): yii\base\Controller->runAction('save-entry', Array)
#16 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/craftcms/cms/src/web/Application.php(246): yii\base\Module->runAction('entries/save-en...', Array)
#17 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/craftcms/cms/src/web/Application.php(385): craft\web\Application->runAction('entries/save-en...', Array)
#18 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/craftcms/cms/src/web/Application.php(206): craft\web\Application->_processActionRequest(Object(craft\web\Request))
#19 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/yiisoft/yii2/base/Application.php(380): craft\web\Application->handleRequest(Object(craft\web\Request))
#20 /Users/mikestecker/Documents/Personal/pbbsite/site/web/index.php(21): yii\base\Application->run()
#21 {main}
```

### Steps to reproduce

1. Create a entry field type, assign to section
2. Create front-end form in a template with this new field, but leave the value blank.

### Additional info

- Craft version: 3.0.0-beta.12
- PHP version: 7.0.17
- Database driver & version: MySQL 5.5.5-10.1.22-MariaDB
- Plugins & versions:
Amazon S3	1.0.3
Doxter	3.1.0
Cookies	1.1.4
Element API	2.0.1
ImageOptim	1.0.6
Minify	1.2.5
PathTools	1.0.4
Patrol	3.0.2
Mailgun	1.0.0.1

-----
Template code:
```twig
<form method="post" accept-charset="UTF-8">
	<input type="hidden" name="action" value="entries/save-entry">
	<input type="hidden" name="sectionId" value="1">
	<input type="hidden" name="enabled" value="1">
	{{ redirectInput('submit/thanks') }}
	{{ csrfInput() }}

	{% macro errorList(errors) %}
		{% if errors %}
		<div class="ui error message">
			<div class="header">We had some issues</div>
			<ul class="list">
				{% for error in errors %}
				<li>{{ error }}</li>
			{% endfor %}
			</ul>
		</div>
		{% endif %}
	{% endmacro %}

	{% from _self import errorList %}


	<div class="field">
		<label for="clientCompanyName">Company Name</label>
		<input type="text" name="fields[clientCompanyName]">
	</div>


	<div class="field">
		<input type="hidden" name="fields[clientEmail]">

		<label for="clientEmail">Client Email</label>
		<div class="matrixblock matrixblock__email email"></div>
		{%- if entry is defined %}
		{% for block in entry.clientEmail.type('email') %}
			<div class="matrixblock email">
				<input type="hidden" name="fields[clientEmail][{{ block.id }}][type]" value="email">
				<input type="hidden" name="fields[clientEmail][{{ block.id }}][enabled]" value="1">

				<input type="text" name="fields[clientEmail][{{ block.id }}][fields][emailAddress]" value="{{ block.emailAddress }}">
				<span><a href="javascript:void(0);" class="remove-email"><i class="fa fa-times"></i> Remove email</a></span>
			</div>
		{% endfor %}
		{% endif %}
		{% if entry is defined %}
			{{ errorList(entry.getErrors('clientEmail')) }}
		{% endif %}
		<button type="button" id="addEmail">Add Email</button>
	</div>


	<div class="field">
		<input type="hidden" name="fields[phone]">

		<label for="phone">Client Phone</label>
		<div class="matrixblock matrixblock__phone phone"></div>
		{%- if entry is defined %}
		{% for block in entry.phone.type('phoneNumber') %}
			<div class="matrixblock phone">
				<input type="hidden" name="fields[phone][{{ block.id }}][type]" value="phoneNumber">
				<input type="hidden" name="fields[phone][{{ block.id }}][enabled]" value="1">

				<input type="text" name="fields[phone][{{ block.id }}][fields][number]" value="{{ block.number }}">
				<span><a href="javascript:void(0);" class="remove-phone"><i class="fa fa-times"></i> Remove phone</a></span>
			</div>
		{% endfor %}
		{% endif %}
		{% if entry is defined %}
			{{ errorList(entry.getErrors('phone')) }}
		{% endif %}
		<button type="button" id="addNumber">Add Number</button>
	</div>

	<div class="field">
		<label for="city">City</label>
		<select class="ui search dropdown" name="fields[city][]" id="citySel">
			<option value="" selected="selected">Please select country first</option>
		</select>
		{% if entry is defined %}
			{{ errorList(entry.getErrors('city')) }}
		{% endif %}
	</div>

	<p>hard-coding these date fields for testing at the moment:</p>
	<input type="text" size="10" name="fields[dateOfVisit][date]" value="{{ now|date('n/j/y') }}">
	<input type="text" size="10" name="fields[dateOfVisit][timezone]" value="America/Los_Angeles">
	{% if entry is defined %}
		{{ errorList(entry.getErrors('dateOfVisit')) }}
	{% endif %}

	<input type="hidden" name="fields[client]" value="">
	<input type="hidden" name="fields[adminNotes]" value="">
	{% if entry is defined %}
		{{ errorList(entry.getErrors('client')) }}
		{{ errorList(entry.getErrors('adminNotes')) }}
	{% endif %}
	<button class="ui button" type="submit">Submit</button>
</form>
```

Here's the javascript for the new matrix fields and the country/city fields (these are category fields):
```js
var countryObject = {
{% for country in countries %}
	"{{ country.title }}": {
		slug: "{{ country.slug }}",
		id: "{{ country.id }}",
		cities: {
{% for city in craft.categories.group('cities').relatedTo(country).orderBy('title asc') %}
			"{{ city.title }}": {
				slug: "{{ city.slug }}",
				id: "{{ city.id }}"
			},
{% endfor %}
		}
	},
{% endfor %}
}

window.onload = function () {
	var countrySel = document.getElementById("countrySel"),
			citySel = document.getElementById("citySel"),
			countryLength = countryObject.length;

	for (var country in countryObject){
		var countryObj = countryObject[country];
		countrySel.options[countrySel.options.length] = new Option(country, countryObj.id);
	}
	countrySel.onchange = function (evt) {
		citySel.length = 1; // remove all options bar first
		if (this.selectedIndex < 1) {
			citySel.options[0].text = "Please select country first"
			return; // done
		}
		citySel.options[0].text = "Please select city";

		var countrySelected = this.value;

		for (var country in countryObject) {
			if(countryObject[country].id == evt.target.value) {
				var citiesObj = countryObject[country].cities;

				for (var city in countryObject[country].cities) {
					var cityObj = countryObject[country].cities[city];
					citySel.options[citySel.options.length] = new Option(city, cityObj.id);
				}
			}
		}

		if (citySel.options.length==2) {
			citySel.selectedIndex=1;
			citySel.onchange();
		}

	}
	countrySel.onchange(); // reset in case page is reloaded
}



var phoneBlockNum = 1;
var emailBlockNum = 1;

/**
 * Event listener to add additional phone numbers(s).
 *
 * @param {object} event | The click event.
 *
 */
$('#addNumber').click(function (event) {
	var markup = '<div class="matrixblock matrixblock__contact phone"><input type="hidden" name="fields[phone][new' + phoneBlockNum + '][type]" value="phoneNumber"><input type="hidden" name="fields[phone][new' + phoneBlockNum + '][enabled]" value="1"><div class="field"><label for="fields[phone][new' + phoneBlockNum + '][fields][number]">Phone Number</label><input type="text" name="fields[phone][new' + phoneBlockNum + '][fields][number]" value=""></div><span><a href="javascript:void(0);" class="remove-phone"><i class="fa fa-times"></i> Remove number</a></span></div>';
	$('div.matrixblock.matrixblock__phone:last').after(markup);
	phoneBlockNum++;
});


/**
 * Event listener to add additional email address(s).
 *
 * @param {object} event | The click event.
 *
 */
$('#addEmail').click(function (event) {
	var markup = '<div class="matrixblock matrixblock__contact email"><input type="hidden" name="fields[clientEmail][new' + emailBlockNum + '][type]" value="email"><input type="hidden" name="fields[clientEmail][new' + emailBlockNum + '][enabled]" value="1"><div class="field"><label for="fields[clientEmail][new' + emailBlockNum + '][fields][emailAddress]">Email</label><input type="text" name="fields[clientEmail][new' + emailBlockNum + '][fields][emailAddress]" value=""></div><span><a href="javascript:void(0);" class="remove-email"><i class="fa fa-times"></i> Remove email</a></span></div>';
	$('div.matrixblock.matrixblock__email:last').after(markup);
	emailBlockNum++;
});


/**
 * Event listener to removing additional phone numbers(s).
 *
 * @param {object} event | The click event.
 *
 */
$('body').on('click', '.remove-phone', function (event) {
	$(this).parent().parent().fadeOut('fast').remove();
	event.preventDefault();
});


/**
 * Event listener to removing additional email address(s).
 *
 * @param {object} event | The click event.
 *
 */
$('body').on('click', '.remove-email', function (event) {
	$(this).parent().parent().fadeOut('fast').remove();
	event.preventDefault();
});
```

**Comment by brandonkelly** [comment]
> The form field I added to the front end for this field is `<input type="hidden" name="fields[client][]" value="">` The field `client` is a entry field to link to another section's entry.

The “blank” input name should be `fields[client]` (no trailing `[]`).

**Comment by mikestecker** [comment]
@brandonkelly I still get the error "Couldn't Save Entry" with the blank input name as you mentioned earlier:
> The “blank” input name should be `fields[client]` (no trailing `[]`).

```
craft\db\QueryAbortedException in /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/craftcms/cms/src/elements/db/ElementQuery.php:854
Stack trace:
#0 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/yiisoft/yii2/db/QueryBuilder.php(92): craft\elements\db\ElementQuery->prepare(Object(craft\db\mysql\QueryBuilder))
#1 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/yiisoft/yii2/db/Query.php(126): yii\db\QueryBuilder->build(Object(craft\elements\db\EntryQuery))
#2 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/yiisoft/yii2/db/Query.php(418): yii\db\Query->createCommand(Object(craft\db\Connection))
#3 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/craftcms/cms/src/db/Query.php(197): yii\db\Query->queryScalar('COUNT(*)', NULL)
#4 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/yiisoft/yii2/db/Query.php(319): craft\db\Query->queryScalar('COUNT(*)', NULL)
#5 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/craftcms/cms/src/elements/db/ElementQuery.php(997): yii\db\Query->count('*', NULL)
#6 [internal function]: craft\elements\db\ElementQuery->count()
#7 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/craftcms/cms/src/validators/ArrayValidator.php(115): count(Object(craft\elements\db\EntryQuery))
#8 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/yiisoft/yii2/validators/Validator.php(267): craft\validators\ArrayValidator->validateAttribute(Object(craft\elements\Entry), 'client')
#9 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/yiisoft/yii2/base/Model.php(358): yii\validators\Validator->validateAttributes(Object(craft\elements\Entry), Array)
#10 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/craftcms/cms/src/services/Elements.php(373): yii\base\Model->validate()
#11 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/craftcms/cms/src/controllers/EntriesController.php(481): craft\services\Elements->saveElement(Object(craft\elements\Entry))
#12 [internal function]: craft\controllers\EntriesController->actionSaveEntry()
#13 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#14 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/yiisoft/yii2/base/Controller.php(156): yii\base\InlineAction->runWithParams(Array)
#15 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/yiisoft/yii2/base/Module.php(523): yii\base\Controller->runAction('save-entry', Array)
#16 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/craftcms/cms/src/web/Application.php(246): yii\base\Module->runAction('entries/save-en...', Array)
#17 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/craftcms/cms/src/web/Application.php(385): craft\web\Application->runAction('entries/save-en...', Array)
#18 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/craftcms/cms/src/web/Application.php(206): craft\web\Application->_processActionRequest(Object(craft\web\Request))
#19 /Users/mikestecker/Documents/Personal/pbbsite/site/vendor/yiisoft/yii2/base/Application.php(380): craft\web\Application->handleRequest(Object(craft\web\Request))
#20 /Users/mikestecker/Documents/Personal/pbbsite/site/web/index.php(21): yii\base\Application->run()
#21 {main}
```

Here are some screenshots showing this as well and you can see the form `POST` as well in the second screenshot.
<img width="924" alt="screenshot 2017-04-18 16 28 52" src="https://cloud.githubusercontent.com/assets/407465/25157155/540a668e-2454-11e7-94b4-6b0759e1a123.png">
<img width="929" alt="screenshot 2017-04-18 16 29 30" src="https://cloud.githubusercontent.com/assets/407465/25157157/58af5488-2454-11e7-979f-a7768171ad25.png">

**Comment by brandonkelly** [comment]
Can you send in a ticket from your Dashboard, including your templates and DB backup, and reference this issue (#1636) and point us to the template the form is on?

**Comment by brandonkelly** [comment]
Fixed via 4da44dcfa218f3c14e1beb3834fde1d765a21a3a and 2e27d40f4757e4a7a1e8d0616982dd49a80cdf36
